File: Two Pointers/M-3. Longest Substring Without Repeating Characters/M-3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  let set = new Set();
  let [l, r] = [0, 0];
  let length = 0;
  let maxLength = 0;

  while (r < s.length) {
    const c = s[r];
    if (!set.has(c)) {
      length++;
      maxLength = Math.max(maxLength, length);
      r++;
      set.add(c);
      continue;
    }
    while (set.has(c)) {
      length--;
      set.delete(s[l++]);
    }
  }

  return maxLength;
};

const testData = [
  {
    s: 'abcabcbb',
    ans: 3,
  },
  {
    s: 'bbbbb',
    ans: 1,
  },
  {
    s: 'pwwkew',
    ans: 3,
  },
];

for (const data of testData) {
  console.log(data.ans === lengthOfLongestSubstring(data.s));
}

File: Two Pointers/M-3. Longest Substring Without Repeating Characters/M-3. Longest Substring Without Repeating Characters.py
class Solution:
    @classmethod
    def lengthOfLongestSubstring(self, s: str):
        map = {}
        ans = 0
        l = 0
        for r in range(len(s)):
            map[s[r]] = (map.get(s[r]) or 0) + 1
            while map[s[r]] > 1:
                map[s[l]] = map[s[l]] - 1
                l += 1
            ans = max(ans, r - l + 1)

        return ans


test_data = [
    {
        "s": 'abcabcbb',
        "ans": 3,
    },
    {
        "s": 'bbbbb',
        "ans": 1,
    },
    {
        "s": 'pwwkew',
        "ans": 3,
    },
]

for data in test_data:
    print(data["ans"] == Solution.lengthOfLongestSubstring(data["s"]))
